R/boost_tree-mboost.R

```diff
@@ -99,12 +99,58 @@ predict_linear_pred._blackboost <- function(object,
   res
 }
 
-# the mboost::survFit isn't able to predict survival probabilities for a given
-# timepoint. This function rounds down to nearest timepoint and uses that
-# for prediction.
-floor_surv_mboost <- function(x, time) {
-  ind <- purrr::map_int(time, ~ max(which(.x > c(-Inf, unname(x$time)))))
-  t(unname(rbind(1, x$surv))[ind, ])
+#' A wrapper for survival probabilities with mboost models
+#' @param x A model from `blackboost()`.
+#' @param new_data Data for prediction.
+#' @param time A vector of integers for prediction times.
+#' @return A tibble with a list column of nested tibbles.
+#' @keywords internal
+#' @export
+#' @examples
+#' library(mboost)
+#' mod <- blackboost(Surv(time, status) ~ ., data = lung, family = CoxPH())
+#' survival_prob_mboost(mod, new_data = lung[1:3, ], time = 300)
+survival_prob_mboost <- function(object, new_data, time) {
+
+  survival_curve <- mboost::survFit(object, newdata = new_data)
+
+  survival_prob <- survival_curve_to_prob(
+    time,
+    event_times = survival_curve$time,
+    survival_prob = survival_curve$surv
+  )
+
+  # survival_prob is length(time) x nrow(new_data) and
+  # `matrix_to_nested_tibbles_survival()` expects the transpose of that (and
+  # then does another t() inside).
+  # this version doesn't need to transpose the matrix at all
+  n_obs <- ncol(survival_prob)
+  ret <- tibble::tibble(
+    .row = rep(seq_len(n_obs), each = length(time)),
+    .time = rep(time, times = n_obs),
+    .pred_survival = as.vector(survival_prob)
+  ) %>%
+    tidyr::nest(.pred = c(-.row)) %>%
+    dplyr::select(-.row)
+
+  ret
+}
+
+survival_curve_to_prob <- function(time, event_times, survival_prob) {
```

Comment:
If this isn't limited to boost, maybe put it in a more general location.

Reply:
Good point. It does work more generally but it's only used by mboost. I'll keep it there for now and rethink moving it at the end of the big refactor!